### ISSUE TITLE
cached runs of dmypy stop reporting "unused type: ignore" warnings

### ISSUE THREAD

**Issue by rggjan** [python/mypy#9655]
Using mypy 0.782, I run
`dmypy run .`

on my codebase. The first time it runs, it correctly reports a few
```
error: unused 'type: ignore' comment
```
same as mypy. However, when rerunning `dmypy run .`, these errors disappear, while all other errors remain.

For some reason, this type of warning (`warn_unused_ignores`) seems to be missed sometimes when using the cache.

Unfortunately, it doesn't seem trivial to reproduce, so I can't give a reproducible project right now...

**Comment by jgillard** [comment]
I noticed this today with mypy 0.942, so this is still an issue.

```shell
> dmypy run
Success: no issues found in 604 source files

> mypy --cache-dir=/dev/null
apps/matchmaking/services.py:260: error: Unused "type: ignore" comment
Found 1 error in 1 file (checked 604 source files)

> mypy --version
mypy 0.942
```

**Comment by peterbygrave** [comment]
This is still a problem in 0.991

**Comment by AlexWaygood** [comment]
Has anybody managed to find a reproducible example yet?

**Comment by meshy** [comment]
I have recently encountered this on 1.0.1. I have not yet produced a minimal example.

In my case the bulk of the `unused type: ignore` warnings which disappear are on lines that have been typed by the `django-stubs` project. I wonder if it may have something to do with that project, or with plugins in general.

**Comment by meshy** [comment]
I have produced a minimal example on Python 3.10.6.

```
# requirements.txt
# Install with `pip install --no-deps -r requirements.txt`

# These are required for mypy
mypy==1.0.1
typing_extensions==4.4.0
mypy-extensions==0.4.3

numpy==1.22.2
```

```python
# example.py
import numpy
a = 1  # type: ignore
```

```
pip install --no-deps -r requirements.txt
dmypy stop
dmypy run -- . --warn-unused-ignores
dmypy run -- . --warn-unused-ignores
```

**Comment by meshy** [comment]
I have a new example that doesn't have any external requirements.

`unused/__init__.py`
(empty file)

`unused/empty.py`
(empty file)

`example.py`
```python
from unused.empty import *
a = 1  # type: ignore
```

Instructions (in a clean virtualenv)

```console
$ dmypy stop
Daemon stopped

$ dmypy run -- --package example --warn-unused-ignores
Daemon started
example.py:2: error: Unused "type: ignore" comment
Found 1 error in 1 file (checked 161 source files)

$ dmypy run -- --package example --warn-unused-ignores
Success: no issues found in 1 source file
```

---

From what I can tell this issue happens when we call `fine_grained_manager.update([], to_delete)` in `dmypy_server.Server.fine_grained_increment_follow_imports`.

In this case, `to_delete` is `[('unused', '/.../mypy-9655/unused/__init__.py')]`.

For some reason, when this update is made, the `Unused "type: ignore" comment` also goes missing.

Next up: dig into `mypy`, and work out what's going on.

**Comment by meshy** [comment]
My [PR now contains a fix](https://github.com/python/mypy/pull/14835) for this that I hope should get things working for most people.

**Comment by ilevkivskyi** [comment]
This should be fixed by https://github.com/python/mypy/pull/14835

**Comment by meshy** [comment]
While I think things have improved, I'm not sure that it was entirely fixed by that PR.

I've produced two more PRs (failing tests only) with similar cases, though I haven't yet been able to get to the bottom of them yet:

- Type ignore comments erroneously marked as unused by dmypy #15043
- More cases showing errors disappearing from dmypy reruns #15049

**Comment by meshy** [comment]
Thanks :)

**Comment by chadrik** [comment]
I've retested with the official 1.3.0 release and I'm still having problems.  In my case, the first run of dmypy produces the correct results, but subsequent runs (after editing a file) produces numerous reports of `Unused "type: ignore" comment`.

**Comment by chadrik** [comment]
Is the above still a known issue or do I need to create a reproduction of the problem?

**Comment by chadrik** [comment]
Scratch that... everything looks to be working with 1.3.0.  thanks!

**Comment by m-honel-ywh2bt** [comment]
Still facing the issue with python 3.12.1 / mypy 1.7.1.

Using [meshy's example](https://github.com/python/mypy/issues/9655#issuecomment-1454087432) in a clean virtualenv:

```
$ dmypy stop
Daemon stopped

$ dmypy --version
dmypy 1.7.1

$ python --version
Python 3.12.1

$ dmypy run -- --package example --warn-unused-ignores
Daemon started
example.py:2: error: Unused "type: ignore" comment  [unused-ignore]
Found 1 error in 1 file (checked 1 source file)

$ dmypy run -- --package example --warn-unused-ignores
Success: no issues found in 1 source file
```

**Comment by meshy** [comment]
@m-honel-ywh2bt we're making progress on this!

Just a couple of days ago we solved the examples in https://github.com/python/mypy/pull/15043 Testing on my local code base suggests that we may have missed some edge-cases there though, so I'm trying to produce some more minimal examples for those.

There are also some cases in https://github.com/python/mypy/pull/15049 that we haven't found the root cause of yet. I hold onto the hope that we'll get to the bottom of this!

**Comment by meshy** [comment]
If anyone following this thread is able to check if https://github.com/python/mypy/pull/15043 works for them, I'd appreciate the feedback.

**Comment by m-honel-ywh2bt** [comment]
@meshy Using https://github.com/python/mypy/pull/15043, I can confirm that `Unused "type: ignore" comment` errors seem to appear everytime dmypy is ran, without having to restart the daemon. 👌🏻

**Comment by JukkaL** [comment]
I think there is another case where type ignore comments still aren't being reported. Here are a few failing test cases:
```
[case testUnusedTypeIgnorePreservedAfterChangeInOtherModule1]
# flags: --warn-unused-ignores
[file main.py]
import a

def f() -> None:
    1()
    a = 1  # type: ignore

[file a.py]
[file a.py.2]
x = 1

[out]
main.py:4: error: "int" not callable
main.py:5: error: Unused "type: ignore" comment
==
main.py:4: error: "int" not callable
main.py:5: error: Unused "type: ignore" comment

[case testUnusedTypeIgnorePreservedAfterChangeInOtherModule2]
# flags: --warn-unused-ignores
[file main.py]
import a

def f() -> None:
    1()
    a = 1  # type: ignore

def g() -> None:
    1()  # type: ignore

[file a.py]
[file a.py.2]
x = 1

[out]
main.py:4: error: "int" not callable
main.py:5: error: Unused "type: ignore" comment
==
main.py:4: error: "int" not callable
main.py:5: error: Unused "type: ignore" comment

```
The error on line 5 isn't reported on the second step.

The root cause is that we don't report unused type ignores in nodes only processed using `reprocess_nodes`, i.e. when following fine-grained dependencies or reprocessing nodes that previously generated errors. 

I think this is a potential way to fix this (but I haven't thought about this very carefully yet):
 * Collect a set of all the triggered/processed nodes in `propagate_changes_using_dependencies`.
 * At the end of a daemon update, perform the equivalent of `generate_unused_ignore_notes`, but only report errors within the previously triggered nodes. The filtering step seems necessary, as some parts of modules are often not processed during a fine-grained increment.

**Comment by meshy** [comment]
@JukkaL good catch.

Do you think this blocks https://github.com/python/mypy/pull/15043, or could we tackle that case separately?

**Comment by JukkaL** [comment]
Ah, I forgot to mention that the test cases I mentioned above aren't specifically covered by #15043 yet. It's okay to leave fixing these for another PR, but we shouldn't close this issue until all known cases are fixed.

**Comment by meshy** [comment]
OK great, thank you. I've removed the "fixes" comment from that PR so we avoid closing this issue.
